core/src/main/scala/cats/ApplicativeError.scala

```diff
@@ -287,6 +287,30 @@ trait ApplicativeError[F[_], E] extends Applicative[F] {
   def catchOnly[T >: Null <: Throwable]: CatchOnlyPartiallyApplied[T, F, E] =
     new CatchOnlyPartiallyApplied[T, F, E](this)
 
+  /**
+   * Often E can be created from Throwable. Here we try to call pure or
+   * catch, adapt into E, and raise.
+   *
+   * Exceptions that cannot be adapted to E will be propagated outside of `F`
+   */
+  def catchNonFatalAs[A](adaptIfPossible: Throwable => Option[E])(a: => A): F[A] =
+    try pure(a)
+    catch {
+      case e if NonFatal(e) => adaptIfPossible(e).map(raiseError[A]).getOrElse(throw e)
+    }
+
+  /**
+   * Evaluates the specified block, catching exceptions of the specified type and maps them to `E`.
+   *
+   * Uncaught exceptions will be propagated outside of `F`
+   *
+   * Note: `catchOnlyAs` assumes that, if a specific exception type `T` is expected, there
+   * exists a mapping from `T` to `E`. If this is not the case, consider either manually
+   * rethrowing inside the mapping function, or using `catchNonFatalAs`
+   */
+  def catchOnlyAs[T >: Null <: Throwable]: CatchOnlyAsPartiallyApplied[T, F, E] =
```

Comment:
Perhaps I'm missing something, but I'm struggling to understand, what the `T >: Null` type boundary is supposed to be for exactly? According to the [Scala Type Hierarchy](https://docs.scala-lang.org/tour/unified-types.html), `Null` automatically inherits to all `AnyRef`/`Object` types. Therefore, for any `T <: Throwable` the `T >: Null` boundary should be met automatically as well. I wonder, are there any examples of cases where `T >: Null` provides any additional benefits?

Reply:
@satorg may be related to https://docs.scala-lang.org/scala3/reference/experimental/explicit-nulls.html

Reply:
Because the `CatchOnlyPartiallyApplied` has both the bound on `Null` and `NotNull` constraint on `T`, I assume there is some type level shenanigans going on that I don't fully understand.

It doesn't _look_ like removing it breaks anything under `++compile`, but there could be nuances I'm missing.

Reply:
@morgen-peschke , I asked in scala-users, just out of curiosity (see this [question](https://discord.com/channels/632150470000902164/632150470000902166/1250597256604684320) and follow up replies).

Turns out, that `T >: Null` helps to prevent from passing `Nothing` explicitly, i.e. `catchOnly[Nothing]`.
It does not seem to be very useful feature, but not completely useless either. So I think let keep it :)